src/Lean/Compiler/LCNF/JoinPoints.lean

```diff
@@ -236,6 +286,235 @@ where
 
 end JoinPointFinder
 
+namespace JoinPointContextExtender
+
+open ScopeM
+
+/--
+The context managed by `ExtendM`.
+-/
+structure ExtendContext where
+  /--
+  The `FVarId` of the current join point if we are currently inside one.
+  -/
+  currentJp? : Option FVarId := none
+  /--
+  The list of valid candidates for extending the context. This will be
+  all `let` and `fun` declarations as well as all `jp` parameters up
+  until the last `fun` declaration in the tree.
+  -/
+  candidates : FVarIdSet := {}
+
+/--
+The state managed by `ExtendM`.
+-/
+structure ExtendState where
+  /--
+  A map from join point `FVarId`s to a respective map from free variables
+  to `Param`s. The free variables in this map are the once that the context
+  of said join point will be extended by by passing in the respective parameter.
+  -/
+  fvarMap : HashMap FVarId (HashMap FVarId Param) := {}
+
+/--
+The monad for the `extendJoinPointContext` pass.
+-/
+abbrev ExtendM := ReaderT ExtendContext StateRefT ExtendState ScopeM
+
+/--
+Replace a free variable if necessary, that is:
+- It is in the list of candidates
+- We are currently within a join point (if we are within a function there
+  cannot be a need to replace them since we dont extend their context)
+- Said join point actually has a replacement parameter registered.
+otherwise just return `fvar`.
+-/
+def replaceFVar (fvar : FVarId) : ExtendM FVarId := do
+  if (←read).candidates.contains fvar then
+    if let some currentJp := (←read).currentJp? then
+      if let some replacement := (←get).fvarMap.find! currentJp |>.find? fvar then
+        return replacement.fvarId
+  return fvar
+
+/--
+Add a new candidate to the current scope + to the list of candidates
+if we are currently within a join point. Then execute `x`.
+-/
+def withNewCandidate (fvar : FVarId) (x : ExtendM α) : ExtendM α := do
+  addToScope fvar
+  if (←read).currentJp?.isSome then
+    withReader (fun ctx => { ctx with candidates := ctx.candidates.insert fvar }) do
+      x
+  else
+    x
+
+/--
+Same as `withNewCandidate` but with multiple `FVarId`s.
+-/
+def withNewCandidates (fvars : Array FVarId) (x : ExtendM α) : ExtendM α := do
+  if (←read).currentJp?.isSome then
+    let candidates := (←read).candidates
+    let folder := (fun acc val => do addToScope val; return acc.insert val)
```

Comment:
Are the parentheses here a workaround for some Lean bug? 
The following should work too.
```lean
let folder acc val := do
  addToScope val
  return acc.insert val
```

Reply:
I didn't even know about that Syntax :D